packages/http/src/validator/validators/body.ts

```diff
@@ -1,82 +1,138 @@
-      const shallowParsedTarget = shallowParseUrlEncoded(target);
-      target = deepParseUrlEncoded(target);
-
-      const encodings = get(content, 'encodings', []);
-      for (const encoding of encodings) {
-        const allowReserved = get(encoding, 'allowReserved', false);
-        const property = encoding.property;
-        const value = shallowParsedTarget[property];
-
-        if (!allowReserved && typeof value === 'string' && value.match(/[\/?#\[\]@!$&'()*+,;=]/)) {
-          return [
-            {
-              path: [prefix, property],
-              message: 'Reserved characters used in request body',
-              severity: DiagnosticSeverity.Error,
-            },
-          ];
-        }
+function validateBody(schema: JSONSchema, target: any) {
+  const diagnostics = validateAgainstSchema(target, schema);
+  return diagnostics.length ? Either.left(diagnostics) : Either.right(target);
+}
 
-        if (encoding.style) {
-          const deserializer = body.get(encoding.style);
-          if (deserializer && schema.properties) {
-            const propertySchema = schema.properties[property];
-            target[property] = deserializer.deserialize(property, target, propertySchema as JSONSchema);
-          }
-        }
-      }
-    }
+function maybeValidateFormBody(
+  schema: JSONSchema,
+  content: IMediaTypeContent,
+  mediaType: string | undefined,
+  target: any,
+) {
+  if (mediaType && typeIs.is(mediaType, 'application/x-www-form-urlencoded')) {
+    const encodings = get(content, 'encodings', []);
+    const encodedUriParams = splitUriParams(target);
 
-    return validateAgainstSchema(target, schema).map(error =>
-      Object.assign({}, error, { path: [prefix, ...(error.path || [])] }),
+    return pipe(
+      validateAgainstReservedCharacters(encodedUriParams, encodings),
+      Either.map(decodeUriEntities),
+      Either.map(deserializeFormBody.bind(undefined, schema, encodings)),
     );
   }
 
-  private getContent(specs: IMediaTypeContent[], mediaType?: string): IMediaTypeContent | undefined {
-    if (!mediaType) {
-      return specs[0];
-    }
+  return Either.right(target);
+}
 
-    const content = specs.find(c => c.mediaType === mediaType);
+function applyPrefix(prefix: string, diagnostics: IPrismDiagnostic[]): IPrismDiagnostic[] {
+  return diagnostics.map(d => ({ ...d, path: [prefix, ...(d.path || [])] }));
+}
 
-    if (!content) {
-      return specs[0];
-    }
+function validateAgainstReservedCharacters(encodedUriParams: Dictionary<string, string>, encodings: IHttpEncoding[]) {
+  return pipe(
+    encodings,
+    Array.reduce<IHttpEncoding, IPrismDiagnostic[]>([], (diagnostics, encoding) => {
+      const allowReserved = get(encoding, 'allowReserved', false);
+      const property = encoding.property;
+      const value = encodedUriParams[property];
+
+      if (!allowReserved && typeof value === 'string' && value.match(/[\/?#\[\]@!$&'()*+,;=]/)) {
```

Comment:
```typescript
if (!allowReserved && typeof value === 'string' && value.match(/[\/?#\[\]@!$&'()*+,;=]/)) {
  return diagnostics.concat({
          path: [property],
          message: 'Reserved characters used in request body',
          severity: DiagnosticSeverity.Error,
        });
} else {
  return diagnostics;
}
```
`.push` does not play well in FP

Reply:
Yeah, but this is simple `Array.reduce`. IMO is exactly the same case as in those two other places (mutable vs immutable). I vote on leaving it as is.

Reply:
if `we are trying to be as FP as possible` (from https://github.com/stoplightio/prism/pull/649/files#r329567219) is to be satisfied, then we go with `.concat`, if not, then `.push` it is